cirq/ops/controlled_gate.py

```diff
@@ -12,52 +12,109 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Any, Union
+from typing import Any, List, Union
 
 import numpy as np
 
 import cirq
 from cirq import linalg, protocols, value
-from cirq.ops import raw_types, controlled_operation as cop
+from cirq.ops import (raw_types, controlled_operation as cop, gate_operation)
 from cirq.type_workarounds import NotImplementedType
 
 
 @value.value_equality
 class ControlledGate(raw_types.Gate):
     """Augments existing gates with a control qubit."""
 
-    def __init__(self, sub_gate: raw_types.Gate) -> None:
+    def __init__(self, sub_gate: raw_types.Gate,
+                 control_qubits: List[raw_types.Qid] = None,
+                 num_controls: int = 1) -> None:
         """Initializes the controlled gate.
 
         Args:
             sub_gate: The gate to add a control qubit to.
+            control_qubits: The qubits that would act as controls.
+            num_control: Total number of control qubits.
         """
-        self.sub_gate = sub_gate
+        if control_qubits is None:
+            control_qubits = []
+        if num_controls < len(control_qubits):
+            raise ValueError('More specified control qubits than num_controls')
+
+        # Leave unspecified controls as Nones.
+        self.control_qubits = ([None]*(num_controls - # type: ignore
+                                       len(control_qubits)) +
+                               control_qubits)
+
+        self.num_controls = num_controls
+
+        # Flatten nested ControlledGates.
+        if isinstance(sub_gate, ControlledGate):
+            self.sub_gate = sub_gate.sub_gate # type: ignore
+            self.control_qubits += sub_gate.control_qubits
+            self.num_controls += sub_gate.num_controls
+        else:
+            self.sub_gate = sub_gate
+
+        self.num_unspecified_controls = self.control_qubits.count(None)
 
     def num_qubits(self) -> int:
-        return self.sub_gate.num_qubits() + 1
+        return self.sub_gate.num_qubits() + self.num_controls
 
     def _decompose_(self, qubits):
         result = protocols.decompose_once_with_qubits(self.sub_gate,
-                                                      qubits[1:],
-                                                      NotImplemented)
+                                            qubits[self.num_controls:],
+                                            NotImplemented)
+
         if result is NotImplemented:
             return NotImplemented
 
-        return [cop.ControlledOperation(qubits[0], op) for op in result]
+        decomposed = []
+        for op in result:
+            controlled_op = op
+            for qubit in qubits[:self.num_controls]:
+                controlled_op = cop.ControlledOperation(qubit, controlled_op)
```

Comment:
Is this the correct order? It looks like this will reverse the order of the qubits.

... I guess it doesn't matter in context.

We may want to generalized `ControlledOperation` at the same time, so we don't need to do this.

Reply:
Reversed all such operations for now. Working on adding the multiqubit ControlledOperation.

Reply:
Done